birdhouse/components/cowbird/docker-compose-extra.yml

```diff
@@ -23,20 +23,25 @@ services:
       # root user
       COWBIRD_FILESYSTEM_ADMIN_UID: 0
       COWBIRD_FILESYSTEM_ADMIN_GID: 0
-      WORKSPACE_DIR: /${USER_WORKSPACES}
+      PUBLIC_WORKSPACE_WPS_OUTPUTS_SUBDIR: ${PUBLIC_WORKSPACE_WPS_OUTPUTS_SUBDIR}
+      SECURE_DATA_PROXY_NAME: ${SECURE_DATA_PROXY_NAME}
+      # Note that WPS_OUTPUTS_DIR and WORKSPACE_DIR must both point to paths from the same volume.
+      # This is to allow the creation of hardlinks between the wpsoutputs and the user workspace.
+      WPS_OUTPUTS_DIR: ${WPS_OUTPUTS_DIR}
+      WORKSPACE_DIR: ${DATA_PERSIST_ROOT}/${USER_WORKSPACES}
     links:
-      - "${COWBIRD_MONGODB_SERVICE}"
+      - cowbird-mongodb
     networks:
       - default
-      - "${COWBIRD_MONGODB_SERVICE}"
+      - cowbird-mongodb
     depends_on:
-      - "${COWBIRD_MONGODB_SERVICE}"
+      - cowbird-mongodb
     volumes:
       - ./components/cowbird/config/cowbird/config.yml:/opt/local/src/cowbird/config/cowbird.yml
       - ./components/cowbird/config/cowbird/cowbird.ini:/opt/local/src/cowbird/config/cowbird.ini
       # even if not running tasks here, they must be registered to send them off to the right place!
       - ./components/cowbird/config/cowbird/celeryconfig.py:/opt/local/src/cowbird/config/celeryconfig.py
-      - "${DATA_PERSIST_ROOT}/${USER_WORKSPACES}:/${USER_WORKSPACES}"
+      - "${DATA_PERSIST_ROOT}:${DATA_PERSIST_ROOT}"
```

Comment:
Woo, this has write-access to the entire `$DATA_PERSIST_ROOT` !!!  Isn't this too much access?

I am guessing it only needs access to `JUPYTERHUB_USER_DATA_DIR="$DATA_PERSIST_ROOT/jupyterhub_user_data"` and Thredds wps_output dir?

Reply:
@ChaamC correct me if I'm wrong... but the reason for this is that both of those directories need to be mounted in the same volume for hardlinks to work: 

@tlvu In #356 we introduce the [`DATA_PERSIST_SHARED_ROOT`](https://github.com/bird-house/birdhouse-deploy/pull/356/files#diff-752de5e561aca0a197c4f5dc44af11b0926716d3f953a1f9fe378670c37d6b10R32) variable here instead. This gives us the ability to mount _only the required subdirectories_ as as single volume.

Reply:
Yes I am a bit behind in most things cowbird.  Understood that for hard links to work, the src and dest of the link should be on the same filesystem.

So if we need to expose Thredds wps_outputs and possibly Geoserver workspace in the Jupyter env workspace, then we need to only volume-mount those 3, in the cowbird container, not the entire `/data`.

If `DATA_PERSIST_SHARED_ROOT` is the same as `DATA_PERSIST_ROOT` by default, then it comes back to the same as exposing the entire `/data` to cowbird writable.

Basically would something like the following works:
```yaml
services:
  cowbird:
    volumes:
      - ./components/cowbird/config/cowbird/config.yml:/opt/local/src/cowbird/config/cowbird.yml
      - ./components/cowbird/config/cowbird/cowbird.ini:/opt/local/src/cowbird/config/cowbird.ini
      # even if not running tasks here, they must be registered to send them off to the right place!
      - ./components/cowbird/config/cowbird/celeryconfig.py:/opt/local/src/cowbird/config/celeryconfig.py
      - "${DATA_PERSIST_SHARED_ROOT}/jupyter_user_data:${DATA_PERSIST_SHARED_ROOT}/jupyter_user_data"
      - "${DATA_PERSIST_SHARED_ROOT}/datasets/wps_outputs:${DATA_PERSIST_SHARED_ROOT}/datasets/wps_outputs"
      - "${DATA_PERSIST_SHARED_ROOT}/geoserver/workspaces:${DATA_PERSIST_SHARED_ROOT}/geoserver/workspaces"
```

So the paths inside and outside of the cowbird container is exactly the same, giving it the impression it has access to the entire `/data` but it's not true.

By the way, I am not sure `${DATA_PERSIST_SHARED_ROOT}/datasets/wps_outputs` actually works since I think the actual wps_outputs is in a data-volume https://github.com/bird-house/birdhouse-deploy/blob/master/birdhouse/config/wps_outputs-volume/docker-compose-extra.yml

Usage of that wps_outputs data-volume:
```
birdhouse/config/raven/config/wps_outputs-volume/docker-compose-extra.yml
6:      - wps_outputs:/data/wpsoutputs

birdhouse/config/thredds/docker-compose-extra.yml
24:      - wps_outputs:/pavics-data/wps_outputs
```

Anyways, many inter-connected pieces so not simple to wrap my head around.

Reply:
> If `DATA_PERSIST_SHARED_ROOT` is the same as `DATA_PERSIST_ROOT` by default, then it comes back to the same as exposing the entire `/data` to cowbird writable.

Yes it would. But... I would like to see `DATA_PERSIST_SHARED_ROOT` not be the same as `DATA_PERSIST_ROOT`. The only reason it is like that by default is to maintain backwards compatibility. 

> Basically would something like the following works:

No, unfortunately that wouldn't work. They actually need to be mounted at the same mount-point. The relative location matters for symlinks, but a shared mount-point matters for hard-links.

> By the way, I am not sure `${DATA_PERSIST_SHARED_ROOT}/datasets/wps_outputs` actually works since I think the actual wps_outputs is in a data-volume

I don't have enough knowledge to comment on this. Sorry

Reply:
> @ChaamC correct me if I'm wrong... but the reason for this is that both of those directories need to be mounted in the same volume for hardlinks to work:

Yes, exactly. If we want to share the wps_outputs data to the user, we need to use hardlinks, and cowbird requires the src and destination of the hardlinks to be in the same volume/file partition, or it will trigger a `Cross-device link`.

I needed to mount the full data directory to be able to make hardlinks between files from the `/data/wps_outputs` and `/data/user_workspaces` directories.

We can update the volume mount to use the upcoming new variable `DATA_PERSIST_SHARED_ROOT` as long as all files used by cowbird for this PR's feature are contained in this directory.

Reply:
Just to double check, for existing deployments, if we want `DATA_PERSIST_SHARED_ROOT=/data_shared_mount` different than `DATA_PERSIST_ROOT=/data`, we have to move 
  * /data/wps_outputs => /data_shared_mount/wps_outputs
  * /data/user_workspaces => /data_shared_mount/user_workspaces
  * anything else needed for the migration?
  
I think this migration should be documented because anyone starting with the current default values and do not want Cowbird to have access to the entire `/data` will have to perform this manual migration.

Or we set a different value immediately in `default.env` and in `env.local.example` we document the backward-compatible value `DATA_PERSIST_SHARED_ROOT=/data` and warn this gives Cowbird too much write-access.  I prefer different default value now to avoid migration later.  Cowbird is not in the default enable list right now so this should not break too many existing deployments.

Also, I still do not understand why we are referring to `/data/wps_outputs` when currently wps_outputs is a data-volume as mentioned in my comment https://github.com/bird-house/birdhouse-deploy/pull/360#discussion_r1336344103 above.  

This means different volume-mount between `wps_outputs` dir and `user_workpaces` dir and hardlink probably do not work.

Does Cowbird fallback to a regular copy when hardlink do not work?  Does it log somewhere when hardlink do not work?

Is there a notebook or test script that tests Cowbird in the PAVICS stack I can try?

Also question for the future as I heard maybe Cowbird will manage Geoserver workspace too.  Does this means we will need to migrate Geoserver data dir under `DATA_PERSIST_SHARED_ROOT` later?

Reply:
> anything else needed for the migration?

That's it, at least concerning this PR's changes.
Although, I'm thinking about something else to watch out for. Not sure if this variable's path could be affected, but changing the [`JUPYTERHUB_USER_DATA_DIR`](https://github.com/bird-house/birdhouse-deploy/blob/547cc085c52924a21d853eeb1149790c2cb7577d/birdhouse/config/jupyterhub/default.env#L22) variable's path could break the symlink that is used in the cowbird workspace to link to the `JUPYTERHUB_USER_DATA_DIR` directory. So, maybe something to watch out for. The user workspaces could eventually be resynchable if we decide to change the jupyterhub path, but the resync is not yet implemented in cowbird for this symlink, so for now it would need a manual recreation of symlinks in each user workspace if that variable is changed.

If we know we will eventually do the migration, I don't see any reason not to do it now. I think it would be okay to change it now if you prefer.

> Also, I still do not understand why we are referring to /data/wps_outputs when currently wps_outputs is a data-volume as mentioned in my comment https://github.com/bird-house/birdhouse-deploy/pull/360#discussion_r1336344103 above.
> This means different volume-mount between wps_outputs dir and user_workpaces dir and hardlink probably do not work.

I am not sure to understand the problem here. Do you mean the fact that the birds who will produce wps_outputs data only mount the wps_outputs volume (and not the user workspace volume)? Any bird that produces wps_outputs only needs access to the wps_outputs volume. Only cowbird requires the full access to the data directory to allow the creation of hardlinks between the wps_outputs and user workspace directories. After that, when the hardlink is created, the volume mount does not require the full data mount to have the hardlink accessible. It is really only for the moment of creation, in cowbird that we require the full mount of the data directory.

> Does Cowbird fallback to a regular copy when hardlink do not work? Does it log somewhere when hardlink do not work?

If the hardlink fails, the failure is only logged in Cowbird's logs. We want to avoid having 2 independant copies that could diverge. We did implement the resync function for the wps_outputs data, so if we need, we can call that API endpoint on cowbird, and cowbird will regenerate any missing hardlinks that could have resulted from a previous failure.

> Is there a notebook or test script that tests Cowbird in the PAVICS stack I can try?

Nope, not at the moment, although I am checking to create a notebook to verify the different parts of the user workspace (notebook symlink, wps_outputs hardlinks, geoserver data, etc.), but I am not sure yet if it is achievable.

> Also question for the future as I heard maybe Cowbird will manage Geoserver workspace too. Does this means we will need to migrate Geoserver data dir under DATA_PERSIST_SHARED_ROOT later?

Cowbird currently already manages geoserver's file permissions on the user workspace. Since the data is only found in the user workspace directly, it doesn't require any symlink/hardlinks, as data is already accessible to the user,it shouldn't be a problem. 
Unless there is external geoserver data to be made accessible to the user that I am not aware of yet?